rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+- support configuring arbitrary storage types in `storage` declarations.
+- support type-checking arbitrary composition of storage types.
+
+The get a hands-on feeling for concrete implementations of the `Storage` trait and better understanding of the meaning of its methods and associated types see the [provided implementations of STD storage types](../files/0013-configurable-and-composable-storage/sway-libs/storage/) and the user-defined [`StoragePair`](../files/0013-configurable-and-composable-storage/user-defined-libs/storage_pair.sw).
+
+Let's explain the `Storage` type by explaining how it is used by the compiler and when implementing storage types.
+
+### Configuring storage
+
+[configuring-storage]: #configuring-storage
+
+The `Value` associated type specifies the type of the _entire_ value that is stored in the storage. Compiler expect the RHS of the _configure with_ operator to be of this type.
+
+E.g., the `StorageBox<T>` stores values of type `T` which means it will define its `Value` as `type Value = T`.
+
+Arbitrary composition of storage types is made simple by intuitive recursive definitions of `Value` associated types.
+
+E.g., the `StorageVec<TStorage>` (note here that every _compound storage type_ can store only types that implement `Storage`) will store slices of whatever the value of the contained `TStorage` is. Thus, its `Value` will simply be `type Value = [TStorage::Value]`.
+
+Similarly, the `StorageMap` will store `(K, V)` tuples, where `V` must be `Storage`. Thus, its `Value` will simply be `type Value = (K, V::Value)`.
+
+Once the `Value` is specified in the `Storage` implementation the compiler can type-check the RHS of `:=` against the `Value` expected by the LHS. It is worth noticing here that we have all the information needed, to, in a case of a type-mismatch, provide a detailed [expressive diagnostics](./0012-expressive-diagnostics.md) explaining the error.
+
+Also, every storage type will always have the same way of configuring, defined by the `Value` type. This guarantees intuitive and consistent usage. E.g., for the `StorageVec`, developers will know that they always need to provide a slice of the values expected by the contained storage type.
+
+The provided RHS expression that returns the type specified by `Value` must be const eval, as it is now.
+
+In the second step, the compiler needs to calculate the concrete storage slots in which to store the _configured with_ RHS values.
+
+This is where the `Storage::Config` associated type comes in play, together with the `const fn` function `internal_get_config()`.
+
+Before generating the slots, the compiler already knows:
+- the `StorageKey` at which a particular `storage` element will be stored. This key is calculated based on the element name and namespace, or it is given using the `in` keyword. This storage key is, as mentioned below, called the _self key_ of the storage element.
```

Comment:
The name of "self key" is an artifact of our thought process and it's littered all over this proposal, but I think we should try to come up with a better one since it's no longer involved with `self` as a keyword.

Maybe "field key"? Since it's the `StorageKey` for a given `storage` field?

Reply:
Well, `self_key` _is_ the best name :smile: IMO, of course.

It very precisely communicates what it is, and that's why it is introduced and consistently used in the RFC:

> An important term to establish will be the _self key_. Self key is the `StorageKey` at which an instance of a storage type is stored.

Why I see it having properties of a perfectly good name:
- In the implementations of `Storage` types, the term `self_key` immediately communicates that "this is _my_ storage key, the storage key at which _I_ am stored".
- The public `self_key()` method is equally clear. It gives a clear answer to the question "what is _your_ key?". (We need this method for, e.g., referencing or any other scenario where the knowledge of the slot location of the storage type instance is needed.)
- In the documentation and communication, being reflexive it immediately points to the storage instance we are talking about. 

Any other word that does not imply reflexive relationship (like "self", or "me") will be to broad and possibly applicable to several instances and the question "whose, of what" will not be immediately incorporated in the word.

E.g., having the `field_key` everywhere rises the question "which field? whose field?". The sentence "Field key is the `StorageKey` at which the instance of a storage type is stored" calls for explaining why the word "field". We have to keep in mind that storage type instances can be used independent of the `storage` which makes the term "field" even less understandable and hard to justify.

Also, some storage types could have clashing terminology with the word we choose. E.g., if we chose "element key", it will clash with what an element in the `StorageVec` is.

That's why I am for the name that implies reflexive relationship and see "self" as the perfect candidate. It also corresponds to what "self" is in Sway types. Something that points to, or denotes, myself.